File: modules/react-native-mlkit-barcode-scanning/ios/RNMLKitBarcodeScanningModule.swift
import ExpoModulesCore
import MLKitBarcodeScanning
import RNMLKitCore

// TODO: move this to core?
// Function to reject a promise with a specified message and domain
func rejectPromiseWithMessage(promise: Promise, message: String, domain: String)
{
    promise.reject(
        NSError(
            domain: domain, code: 1,
            userInfo: [NSLocalizedDescriptionKey: message])
    )
}

let ERROR_DOMAIN: String =
    "red.infinite.reactnativemlkit.BarcodeScanningErrorDomain"

public class RNMLKitBarcodeScanningModule: Module {
    var barcodeScanner: BarcodeScanner? = nil

    public func definition() -> ModuleDefinition {
        Name("RNMLKitBarcodeScanning")

        AsyncFunction("initialize") { (promise: Promise) in
            self.barcodeScanner = BarcodeScanner.barcodeScanner()
            promise.resolve(nil)
        }

        AsyncFunction("process") { (imagePath: String, promise: Promise) in
            let logger = Logger(logHandlers: [
                createOSLogHandler(category: Logger.EXPO_LOG_CATEGORY)
            ])

            Task {
                do {
                    guard let barcodeScanner = self.barcodeScanner else {
                        rejectPromiseWithMessage(
                            promise: promise,
                            message:
                                "[RNMLKitBarcodeScanning.process] Barcode Scanner not initiliazed",
                            domain: ERROR_DOMAIN)
                        return
                    }
                    let image = try RNMLKitImage(imagePath: imagePath)
                    let barcodes = try await barcodeScanner.process(image.visionImage)

                    logger.debug(barcodes)

                    let result = RNMLKitBarcodeScannerResult(barcodes: barcodes)
                    promise.resolve(result.record)
                } catch {
                    rejectPromiseWithMessage(promise: promise, message: "[RNMLKitBarcodeScanning.process] Error processing barcode: \(error)", domain: ERROR_DOMAIN)
                }
            }
        }
    }
}

File: modules/react-native-mlkit-barcode-scanning/android/src/main/java/red/infinite/reactnativemlkit/barcodescanning/RNMLKitBarcodeScanningModule.kt
package red.infinite.reactnativemlkit.barcodescanning

import android.net.Uri

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.runBlocking
import red.infinite.reactnativemlkit.core.RNMLKitModule

class RNMLKitBarcodeScanningModule : RNMLKitModule("RNMLKitBarcodeScanning") {
  private var barcodeScanner: RNMLKitBarcodeScanner? = null
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('RNMLKitBarcodeScanning')` in JavaScript.
    Name("RNMLKitBarcodeScanning")


    AsyncFunction("initialize") { options: RNMLKitBarcodeScannerOptionsRecord, promise: Promise ->
        log.d("initialize: Initializing Barcode Scanner")
        val scannerOptions = RNMLKitBarcodeScannerOptions(options)
        barcodeScanner = RNMLKitBarcodeScanner(scannerOptions)
        log.d("initialize: Barcode Scanner initialized")
        promise.resolve(null)
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("process") { imagePath: String, promise: Promise ->
      val imageUri = Uri.parse(imagePath)

      runBlocking suspend@{
        val barcodeScanner = barcodeScanner
        if (barcodeScanner == null) {
          handleException(promise, message = "Barcode Scanner not initialized")
          return@suspend
        }
        try {
          var result =
              barcodeScanner.detectInImage(imagePath, appContext).getOrElse {
                handleException(promise, it)
                throw it
              }
          promise.resolve(result.record)
        } catch (e: Exception) {
          handleException(promise, e)
        }

      }
    }
  }
}
